File: src/extract-pdf-pages/extract-pdf-pages.mjs
#!/usr/bin/env node
// Recursively extract text from PDFs using pdf-parse (pdf.js under the hood).
// Usage:
//   node scripts/extract-pdf-text.mjs <input_dir> [output_dir]
// Install dependency first:
//   pnpm add pdf-parse
// Falls back with a clear error if pdf-parse is missing.

import { promises as fs } from "fs";
import path from "path";
import process from "process";

let pdfParse;
try {
  ({ default: pdfParse } = await import("pdf-parse"));
} catch (e) {
  console.error("Dependency pdf-parse not installed. Run: pnpm add pdf-parse");
  process.exit(1);
}

if (process.argv.length < 3) {
  console.error(
    "Usage: node scripts/extract-pdf-text.mjs <input_dir> [output_dir]"
  );
  process.exit(1);
}

const inputDir = path.resolve(process.argv[2]);
const outputDir = path.resolve(
  process.argv[3] || path.join(inputDir, "text-output-node")
);

async function listPdfFiles(dir) {
  const entries = await fs.readdir(dir, { withFileTypes: true });
  const files = [];
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      files.push(...(await listPdfFiles(fullPath)));
    } else if (entry.isFile() && entry.name.toLowerCase().endsWith(".pdf")) {
      files.push(fullPath);
    }
  }
  return files;
}

async function extractPdf(pdfPath) {
  const data = await fs.readFile(pdfPath);
  const result = await pdfParse(data); // result.text
  return result.text; // Plain reflowed text
}

async function main() {
  try {
    await fs.mkdir(outputDir, { recursive: true });
    const pdfs = await listPdfFiles(inputDir);
    if (!pdfs.length) {
      console.log("No PDF files found.");
      return;
    }
    console.log(`Found ${pdfs.length} PDF files`);
    for (const pdf of pdfs) {
      const rel = path.relative(inputDir, pdf);
      const outPath = path.join(outputDir, rel.replace(/\.pdf$/i, ".txt"));
      await fs.mkdir(path.dirname(outPath), { recursive: true });
      try {
        const text = await extractPdf(pdf);
        await fs.writeFile(outPath, text, "utf8");
        console.log(`OK  ${rel}`);
      } catch (e) {
        console.error(`ERR ${rel}: ${e.message}`);
      }
    }
    console.log("Done.");
  } catch (e) {
    console.error("Fatal:", e);
    process.exit(1);
  }
}

main();

File: src/extract-pdf-pages/extract-pdf-pages.py
#!/usr/bin/env python3
"""
Extract the first N pages from multiple PDF files.

Usage:
    python extract-pdf-pages.py <input_directory> <num_pages> [output_directory] [--skip-first]

Example:
    python extract-pdf-pages.py ./pdfs 3
    python extract-pdf-pages.py ./pdfs 5 ./extracted
    python extract-pdf-pages.py ./pdfs 4 ./extracted --skip-first
"""

import sys
from pathlib import Path

try:
    from pypdf import PdfReader, PdfWriter
except ImportError:
    print("Error: pypdf not installed. Install with: pip install pypdf")
    sys.exit(1)


def extract_first_n_pages(input_path, output_path, num_pages, skip_first=False):
    """Extract the first N pages from a PDF file."""
    try:
        reader = PdfReader(input_path)
        writer = PdfWriter()

        # Calculate start page and pages to extract
        start_page = 1 if skip_first else 0
        end_page = start_page + num_pages
        pages_to_extract = min(end_page, len(reader.pages))

        for page_num in range(start_page, pages_to_extract):
            writer.add_page(reader.pages[page_num])

        with open(output_path, "wb") as output_file:
            writer.write(output_file)

        return True, pages_to_extract
    except Exception as e:
        return False, str(e)


def main():
    if len(sys.argv) < 3:
        print(__doc__)
        sys.exit(1)

    # Parse arguments
    skip_first = "--skip-first" in sys.argv
    args = [arg for arg in sys.argv[1:] if arg != "--skip-first"]

    input_dir = Path(args[0])
    num_pages = int(args[1])
    output_dir = Path(args[2]) if len(args) > 2 else input_dir / "extracted"

    if not input_dir.exists():
        print(f"Error: Input directory '{input_dir}' does not exist")
        sys.exit(1)

    if not 1 <= num_pages <= 5:
        print("Warning: num_pages should be between 1 and 5")

    # Create output directory if it doesn't exist
    output_dir.mkdir(parents=True, exist_ok=True)

    # Find all PDF files recursively in nested folders
    pdf_files = list(input_dir.glob("**/*.pdf"))

    if not pdf_files:
        print(f"No PDF files found in '{input_dir}'")
        sys.exit(0)

    print(f"Found {len(pdf_files)} PDF files")
    if skip_first:
        print(f"Extracting pages 2-{num_pages + 1} from each (skipping first page)...")
    else:
        print(f"Extracting first {num_pages} pages from each...")
    print(f"Output directory: {output_dir}\n")

    success_count = 0
    error_count = 0

    for pdf_path in pdf_files:
        if skip_first:
            output_filename = f"{pdf_path.stem}_pages_2_to_{num_pages + 1}.pdf"
        else:
            output_filename = f"{pdf_path.stem}_first_{num_pages}_pages.pdf"
        output_path = output_dir / output_filename

        success, result = extract_first_n_pages(
            pdf_path, output_path, num_pages, skip_first
        )

        if success:
            print(f"✓ {pdf_path.name} -> {output_filename} ({result} pages)")
            success_count += 1
        else:
            print(f"✗ {pdf_path.name} - Error: {result}")
            error_count += 1

    print(f"\nComplete: {success_count} successful, {error_count} errors")


if __name__ == "__main__":
    main()
